File: runtimes/org.eclipse.edt.runtime.java/src/org/eclipse/edt/runtime/java/egl/lang/AnyException.java
/*******************************************************************************
 * Copyright © 2006, 2011 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * IBM Corporation - initial API and implementation
 *
 *******************************************************************************/
package org.eclipse.edt.runtime.java.egl.lang;

import java.lang.reflect.Field;

import org.eclipse.edt.javart.Constants;
import org.eclipse.edt.javart.JavartException;
import org.eclipse.edt.javart.TypeConstraints;

public class AnyException extends JavartException implements egl.lang.AnyException {
	private static final long serialVersionUID = Constants.SERIAL_VERSION_UID;

	public String message;

	public AnyException() {
		super();
	}

	public AnyException(String id) {
		super(id);
	}

	public AnyException(String id, String message) {
		super(id, message);
		this.message = message;
		ezeInitialize();
	}

	public AnyException(Throwable ex) {
		super(ex);
		ezeInitialize();
	}

	public Object clone() throws CloneNotSupportedException {
		AnyException ezeClone = (AnyException) super.clone();
		return ezeClone;
	}

	@Override
	public void ezeInitialize() {}

	@Override
	public egl.lang.EglAny ezeGet(String name) throws JavartException {
		try {
			Field field = this.getClass().getField(name);
			Object value = field.get(this);
			return value instanceof egl.lang.EglAny ? (egl.lang.EglAny) value : EglAny.ezeBox(value);
		}
		catch (Exception e) {
			throw new DynamicAccessException(e);
		}
	}

	@Override
	public String ezeName() {
		return this.getClass().getSimpleName();
	}

	@Override
	public void ezeSet(String name, Object value) throws JavartException {
		try {
			Field field = this.getClass().getField(name);
			field.set(this, value);
		}
		catch (Exception e) {
			throw new DynamicAccessException(e);
		}
	}

	@Override
	public TypeConstraints ezeTypeConstraints(String fieldName) {
		return null;
	}

	@Override
	public String ezeTypeSignature() {
		return this.getClass().getName();
	}

	public String toString() {
		String msg = getLocalizedMessage();
		if (msg == null) {
			return ezeTypeSignature();
		} else {
			String typeSig = ezeTypeSignature();
			return new StringBuilder(typeSig + 1 + msg.length()).append(typeSig).append(' ').append(msg).toString();
		}
	}

	@Override
	public void setMessage(String message) {
		this.message = message;

	}

	@Override
	public String getMessage() {
		return message == null ? super.getMessage() : message;
	}

	@Override
	public AnyException ezeUnbox() {
		return this;
	}

}

File: runtimes/org.eclipse.edt.runtime.javascript/runtime/eglx/lang/AnyException.js
/*******************************************************************************
 * Copyright © 2008, 2013 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * IBM Corporation - initial API and implementation
 *
 *******************************************************************************/

//factory for AnyException
//required because AnyException is a subclass of Record
//agruments for Record is an array of pairs, representing fields
egl.createAnyException = function( /*string*/ messageID, /*string or array*/ inserts )
{
	if (typeof(inserts) != "string") {
		inserts = egl.getRuntimeMessage( messageID, inserts );
	}

	var args = new Array();
	args.push( [ "messageID", messageID || "" ] );
	args.push( [ "message", inserts || "" ] );
	// uncomment the following line to see exceptions being swallowed
	// egl.println("Exception: "+inserts+" "+this);
	return new egl.eglx.lang.AnyException( args );
};

egl.defineClass( "eglx.lang", "AnyException", "egl.jsrt", "Record", {
	
	//to create an exception in the runtime:
	//	var messageID = "EGLXXXX";
	//	var msgInserts = [ insert1, insert2, ...];
	//	var ex = egl.createNameOfException( messageID, msgInserts );
	//	throw ex;
	//
	//or, more simply:
	//	throw egl.createNameOfException( "EGLXXXX", [ insert1, insert2, ...], value 
	//where "value" is an exception-specific field. There can be more than one value.
	//
	//if you already have the message and don't want to look it up:
	//	throw new egl.createNameOfException( "EGLXXXX", "message" );
	//the second argument is differentiated as being a string, not an array
	//
	//any arguments are optional, since they set exception-specific data
	
	
	"constructor" : function( messageID, inserts ) {
		//subclass of Record
		//call superclass constructor
		this.eze$$isNullable = true;
		this.messageID = this.messageID || "";
		this.message = this.message || "";
	},
	
	"getValue" : function() {
		return this;
	},
	
	"toString" : function() {
		return this.message;
	},
	
	"toJSONString" : function() {
		var s = "{";
		var i = 0;
		for (var f in this) {
			if ( typeof f == "function" ) continue; //ignore functions
			if (i>0) s += ",";
			i++;
			s += "\"" + f + "\":" + this[f];
		}
		return s + "}";
	},
	
	"eze$$getName": function() {
		return "AnyException";
	},
	
	"eze$$getChildVariables": function() {
		var eze$$parent = this;
		return [
		{name: "message", value : eze$$parent.message, type : "eglx.lang.EString", jsName : "message"},
		{name: "messageID", value : eze$$parent.messageID, type : "eglx.lang.EString", jsName : "messageID"}
		];
	}
	
});
